Improve readability and completedness for Redis (#9093)
nextcloud/documentation: admin_manual/configuration_server/caching_configuration.rst

* `Memcached <https://www.memcached.org/>`_
   For distributed caching.
   
Data caches, or memcaches, must be explicitly configured in Nextcloud by installing
and enabling your desired cache, and then adding the appropriate entry to 
``config.php`` (See :doc:`config_sample_php_parameters` for an overview of
all possible config parameters).

Recommendations based on type of deployment
-------------------------------------------

You may use both a local and a distributed cache. Recommended caches are APCu 
and Redis. After installing and enabling your chosen memcache (data cache),
verify that it is active by running :ref:`label-phpinfo`.

.. note:: See specific cache configuration options under the appropriate section further down.

Small/Private home server
^^^^^^^^^^^^^^^^^^^^^^^^^

Only use APCu::

>     'memcache.local' => '\OC\Memcache\APCu',

Reviewer: also set distributed, because both may be used, and the use case is a different one.

Author: @blizzz also this one has been there [before](https://github.com/nextcloud/documentation/pull/9093/files#diff-48dc187eeb03371b5ad6886b33043bf060884809c1d51a3568c5d29e40e3bf13L221-L227).

  'memcache.local' => '\OC\Memcache\APCu',
  'memcache.distributed' => '\OC\Memcache\Redis',
  'memcache.locking' => '\OC\Memcache\Redis',
  'redis.cluster' => [
      'seeds' => [ // provide some/all of the cluster servers to bootstrap discovery, port required
         'cache-cluster:7000',
         'cache-cluster:7001',
      ],
   ]

... or Memcached cluster ...::

  'memcache.local' => '\OC\Memcache\APCu',
  'memcache.distributed' => '\OC\Memcache\Memcached',
  'memcache.locking' => '\OC\Memcache\Memcached',
  'memcached_servers' => [
      [ 'server1.example.com', 11211 ],
      [ 'server2.example.com', 11211 ],
   ],

... for distrebuted and locking caches.

.. note:: If you run multiple web servers and enable a distributed cache in
    your ``config.php`` (``memcache.distributed``) or a file locking provider
    (``memcache.locking``) you need to make sure that they are referring to the
    very same memcache server/cluster and not to ``localhost`` or a unix socket.

Additional notes for Redis vs. APCu on memory caching
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

APCu is faster at local caching than Redis. If you have enough memory, use APCu for Memory Caching

> and Redis for File Locking. If you are low on memory, use Redis for both.

Reviewer: That probably opens up questions on "why" and also wat "low memory" actually means.

Author: This is not new, I've just moved it to a different place: see [current doc](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/caching_configuration.html#additional-notes-for-redis-vs-apcu-on-memory-caching)

Author: Also here. I think it would be great, if we could go ahead with this PR, or assign someone who is capable to technically review the content. Something which _I_ am not capable of. What do you say, @blizzz?

 22203 ? Ssl    0:00 /usr/bin/redis-server 127.0.0.1:6379 
 
Restart your Web server, add the appropriate entries to your ``config.php``, and 
refresh your Nextcloud admin page.

Redis configuration in Nextcloud (config.php)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For best performance, use Redis for file locking by adding this::

  'memcache.locking' => '\OC\Memcache\Redis',

Further more, you can use Redis for the distributed server cache:: 

>   'memcache.distributed' => '\OC\Memcache\Redis',

Reviewer: It's not furthermore, but the essential configuration

Author: Why essential? You could use _memcached_ for the distributed cache instead (`'memcache.distributed' => '\OC\Memcache\Memcached'`), and _Redis_ only for local cache. As such, _Redis_ is just one available option for the distributed cache. That's why I used "further more".

Author: After reading your next comment, it appears that using Redis for local cache only doesn't make much sense. I wonder if that would not need to be explained in some part of the document. I.e. recommended setups are:

- APCu for local cache
- Redis for distributed cache and file locking on a single server
- Redis cluster or Memcached cluster for muliti-server setups

Or something along those lines.

Reviewer: > Why essential? You could use _memcached_ for the distributed cache instead (`'memcache.distributed' => '\OC\Memcache\Memcached'`), and _Redis_ only for local cache. As such, _Redis_ is just one available option for the distributed cache. That's why I used "further more".

For me it sounded that the `'memcache.distributed'` option is a nice to have.

Additionally, you can use Redis for the local cache like so:: 

>   'memcache.local' => '\OC\Memcache\Redis',

Reviewer: You can do it on single-server systems only, with more than app servers it can cause trouble. Even on single-server setups APCu should be faster. So this is possible, but not recommendable.

Author: I've incorporated your comments in a note.

 22203 ? Ssl    0:00 /usr/bin/redis-server 127.0.0.1:6379 
 
Restart your Web server, add the appropriate entries to your ``config.php``, and 
refresh your Nextcloud admin page.

Redis configuration in Nextcloud (config.php)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For best performance, use Redis for file locking by adding this::

  'memcache.locking' => '\OC\Memcache\Redis',

> Further more, you can use Redis for the distributed server cache:: 
> 
>   'memcache.distributed' => '\OC\Memcache\Redis',

Reviewer 2: ```suggestion
Additionally, you should use Redis for the distributed server cache:: 

  'memcache.distributed' => '\OC\Memcache\Redis',
```

   'memcache.distributed' => '\OC\Memcache\Redis',
   'memcache.local' =>'\OC\Memcache\Redis' ,
   'redis' => [
      // 'host'      => see connection parameters below
      // 'port'      => see connection parameters below
     'user'          => 'nextcloud',
     'password'      => 'password',
     'dbindex'       => 0,
     'timeout'       => 1.5,
     'read_timeout'  => 1.5,
   ]

The following options are available to configure when using a redis cluster (all but ``seeds`` are optional)::

   'memcache.locking' => '\OC\Memcache\Redis',
   'memcache.distributed' => '\OC\Memcache\Redis',
   'memcache.local' =>'\OC\Memcache\Redis' ,
   'redis.cluster' => [
      'seeds' => [ // provide some/all of the cluster servers to bootstrap discovery, port required
         'cache-cluster:7000',
         'cache-cluster:7001',
         'cache-cluster:7002',
         'cache-cluster:7003',
         'cache-cluster:7004',
         'cache-cluster:7005'
      ],
      'failover_mode'   => \RedisCluster::FAILOVER_ERROR
      'timeout'         => 0.0,
      'read_timeout'    => 0.0,
      'user'            => 'nextcloud',
      'password'        => 'password',
      'dbindex'         => 0,
   ]

> Connecting to Single Redis Server over TCP

Reviewer 3: ```suggestion
Connecting to single Redis server over TCP
```

More information on configuration of phpredis session handler can be found on the
`PhpRedis GitHub page <https://github.com/phpredis/phpredis>`_

Additional Redis installation help
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If your version of **Mint** or **Ubuntu** does not package the required version of 
``php-redis``, then try `this Redis guide on Tech and Me
<https://www.techandme.se/install-redis-cache-on-ubuntu-server-with-php-7-and-nextcloud/>`_
for a complete Redis installation on Ubuntu 14.04 using PECL. 
These instructions are adaptable for **any distro** that does not package the 
supported version, or that does not package Redis at all, such as **SUSE Linux 
Enterprise Server** and **Red Hat Enterprise Linux**.

> For PHP 7.0 and PHP 7.1 use Redis PHP module 3.1.x or later.

Reviewer: I think we can spare us this section, for 7.4 is the minimum requirement at the moment.

Author: This is not my stuff. OF course it can be removed, but it has been there [before](ation/pull/9093/files#diff-48dc187eeb03371b5ad6886b33043bf060884809c1d51a3568c5d29e40e3bf13L249).

If you want me to remove it, please specify exactly which part (lines). But it is a shame, if the improvements of this PR are solely delayed due to old stuff. And honestly, I do not know enough about these technical details to be able to "write" the manual. I've only re-organized and added more examples and explanations ... :-)
